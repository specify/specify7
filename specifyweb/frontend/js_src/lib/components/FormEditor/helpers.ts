import type { RR } from '../../utils/types';
import type { Tables } from '../DataModel/types';

export const tablesWithBusRulesIn6 = new Set([
  'AccessionAgentBusRules',
  'AccessionAuthorizationBusRules',
  'AccessionBusRules',
  'AddressBusRules',
  'AgentBusRules',
  'AppraisalBusRules',
  'AttachmentBusRules',
  'AuthorBusRules',
  'BioStratBusRules',
  'BorrowBusRules',
  'CatalogNumberingSchemeBusRules',
  'CatAutoNumberingSchemeBusRules',
  'CollectingEventBusRules',
  'CollectingTripAuthorizationBusRules',
  'CollectingTripBusRules',
  'CollectionBusRules',
  'CollectionObjectBusRules',
  'CollectorBusRules',
  'ConservDescriptionBusRules',
  'ConservEventBusRules',
  'ContainerBusRules',
  'DataTypeBusRules',
  'DeaccessionAgentBusRules',
  'DeterminationBusRules',
  'DeterminerBusRules',
  'DisciplineBusRules',
  'DivisionBusRules',
  'DNASequenceBusRules',
  'DNASequencingRunBusRules',
  'ExchangeOutBusRules',
  'ExchangeOutPrepBusRules',
  'ExtractorBusRules',
  'FieldNotebookBusRules',
  'FieldNotebookPageSetBusRules',
  'FundingAgentBusRules',
  'GeographyBusRules',
  'GeologicTimePeriodBusRules',
  'GiftBusRules',
  'GiftPreparationBusRules',
  'GroupPersonBusRules',
  'InfoRequestBusRules',
  'LithoStratBusRules',
  'LoanBusRules',
  'LoanGiftShipmentBusRules',
  'LoanPreparationBusRules',
  'LoanReturnPreparationBusRules',
  'LocalityBusRules',
  'PcrPersonBusRules',
  'PermitBusRules',
  'PickListBusRules',
  'PreparationBusRules',
  'PrepTypeBusRules',
  'SpecifyUserBusRules',
  'StorageBusRules',
  'TaxonBusRules',
  'TreatmentEventBusRules',
  'TreeDefBusRules',
]);

/**
 * Most of the time business rules class name can be inferred from table name.
 * Exceptions:
 */
export const businessRulesOverride: Partial<RR<keyof Tables, string>> = {
  Shipment: 'LoanGiftShipment',
  CollectionObjectAttachment: 'Attachment',
};

export const getBusinessRuleClassFromTable = (
  tableName: keyof Tables
): string =>
  businessRulesOverride[tableName] === undefined
    ? tablesWithBusRulesIn6.has(`${tableName}BusRules`)
      ? `edu.ku.brc.specify.datamodel.busrules.${tableName}BusRules`
      : ''
    : `edu.ku.brc.specify.datamodel.busrules.${businessRulesOverride[
        tableName
      ]!}BusRules`;
