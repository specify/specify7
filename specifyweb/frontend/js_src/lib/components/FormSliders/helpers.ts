import { Relationship } from '../DataModel/specifyField';

export const shouldBeToOne = (
  relationship: Relationship | undefined
): boolean =>
  relationship?.type.includes('-to-many') === false ||
  (relationship?.type.includes('-to-many') === true &&
    !relationship.isDependent() &&
    relationship.getReverse()?.isDependent() === true);
