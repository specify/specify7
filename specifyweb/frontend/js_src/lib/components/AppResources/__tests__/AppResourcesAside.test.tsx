import React from 'react';
import * as Router from 'react-router-dom';

import { requireContext } from '../../../tests/helpers';
import { mount } from '../../../tests/reactUtils';
import type { RA } from '../../../utils/types';
import type { AppResourcesConformation } from '../Aside';
import { AppResourcesAside } from '../Aside';
import { testAppResources } from './testAppResources';

requireContext();

describe('AppResourcesAside (simple no conformation case)', () => {
  test('simple no conformation case', () => {
    const onOpen = jest.fn();
    const setConformations = jest.fn();

    const { asFragment, unmount } = mount(
      <AppResourcesAside
        conformations={[[], setConformations]}
        filters={undefined}
        isEmbedded
        resources={testAppResources}
        onOpen={onOpen}
      />
    );

    expect(asFragment()).toMatchSnapshot();
    unmount();
  });
});

describe('AppResourcesAside (expanded case)', () => {
  test('expanded case', async () => {
    const onOpen = jest.fn();
    const _setConformations = jest.fn();
    let _conformations: RA<AppResourcesConformation> | undefined = [];

    const setConformations = (
      argument: typeof _conformations // Whatever
    ) => {
      _setConformations(argument);
      _conformations = argument;
    };

    const {
      getAllByRole,
      user,
      unmount,
      asFragment: asFragmentInitial,
    } = mount(
      <AppResourcesAside
        conformations={[_conformations, setConformations]}
        filters={undefined}
        isEmbedded
        resources={testAppResources}
        onOpen={onOpen}
      />
    );

    const initialFragment = asFragmentInitial().textContent;

    const buttons = getAllByRole('button');

    await user.click(buttons[1]);

    unmount();

    const {
      asFragment,
      unmount: unmountSecond,
      getAllByRole: getIntermediate,
    } = mount(
      <AppResourcesAside
        conformations={[_conformations, setConformations]}
        filters={undefined}
        isEmbedded
        resources={testAppResources}
        onOpen={onOpen}
      />
    );

    expect(asFragment()).toMatchSnapshot();

    const intermediateFragment = asFragment().textContent;

    const closeAllButton = getIntermediate('button').at(-1);
    await user.click(closeAllButton!);

    unmountSecond();

    const {
      asFragment: asFragmentLater,
      unmount: unmountThird,
      getAllByRole: getFinal,
    } = mount(
      <AppResourcesAside
        conformations={[_conformations, setConformations]}
        filters={undefined}
        isEmbedded
        resources={testAppResources}
        onOpen={onOpen}
      />
    );

    const laterFragment = asFragmentLater().textContent;

    expect(initialFragment).toBe(
      'Global Resources (2)Discipline Resources (4)Expand AllCollapse All'
    );
    expect(intermediateFragment).toBe(
      'Global Resources (2)Discipline Resources (4)Botany (4)Expand AllCollapse All'
    );
    expect(laterFragment).toBe(
      'Global Resources (2)Discipline Resources (4)Expand AllCollapse All'
    );

    const expandAllButton = getFinal('button')[2];
    await user.click(expandAllButton);

    unmountThird();

    const { asFragment: asFragmentAllExpanded, unmount: unmountExpandedll } =
      mount(
        <Router.MemoryRouter initialEntries={['/specify/resources/']}>
          <AppResourcesAside
            conformations={[_conformations, setConformations]}
            filters={undefined}
            isEmbedded
            resources={testAppResources}
            onOpen={onOpen}
          />
        </Router.MemoryRouter>
      );

    const expandedAllFragment = asFragmentAllExpanded().textContent;

    expect(expandedAllFragment).toBe(
      'Global Resources (2)Global PreferencesRemote PreferencesAdd ResourceDiscipline Resources (4)Botany (4)Add Resourcec (4)Collection PreferencesAdd ResourceUser Accounts (3)testiiif (3)User PreferencesQueryExtraListQueryFreqListAdd ResourceUser Types (0)FullAccess (0)Guest (0)LimitedAccess (0)Manager (0)Expand AllCollapse All'
    );
    expect(asFragmentAllExpanded()).toMatchSnapshot();
    unmountExpandedll();
  });
});
