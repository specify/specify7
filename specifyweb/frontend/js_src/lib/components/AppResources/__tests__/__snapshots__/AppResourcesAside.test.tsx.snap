// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppResourcesAside (expanded case) expanded case 1`] = `
<DocumentFragment>
  <aside
    class="
        !gap-2 sm:overflow-visible
        flex flex-col gap-4 overflow-scroll
  overflow-x-auto [overflow-y:overlay] [scrollbar-gutter:auto] 
      "
  >
    <ul
      class=" flex flex-1 flex-col gap-1 overflow-auto"
      role="tree"
    >
      <li
        aria-expanded="false"
        aria-labelledby="app-resources-tree-4-label"
        class="flex flex-col gap-2"
        id="app-resources-tree-4-li"
        role="treeitem"
      >
        <button
          aria-controls="app-resources-tree-4-li"
          class="link inline text-left font-bold"
          id="app-resources-tree-4-label"
          type="button"
        >
          Global Resources 
          <span
            class="pl-2 text-neutral-500"
          >
            (2)
          </span>
        </button>
      </li>
      <li
        aria-expanded="true"
        aria-labelledby="app-resources-tree-5-label"
        class="flex flex-col gap-2"
        id="app-resources-tree-5-li"
        role="treeitem"
      >
        <button
          aria-controls="app-resources-tree-5-li"
          class="link inline text-left font-bold"
          id="app-resources-tree-5-label"
          type="button"
        >
          Discipline Resources 
          <span
            class="pl-2 text-neutral-500"
          >
            (4)
          </span>
        </button>
        <ul
          aria-label="Sub-categories"
          class=" flex flex-col gap-2 pl-4"
          role="group"
        >
          <li
            aria-expanded="false"
            aria-labelledby="app-resources-tree-6-label"
            class="flex flex-col gap-2"
            id="app-resources-tree-6-li"
            role="treeitem"
          >
            <button
              aria-controls="app-resources-tree-6-li"
              class="link inline text-left font-bold"
              id="app-resources-tree-6-label"
              type="button"
            >
              Botany 
              <span
                class="pl-2 text-neutral-500"
              >
                (4)
              </span>
            </button>
          </li>
        </ul>
      </li>
    </ul>
    <div
      class="flex flex-wrap gap-2"
    >
      <button
        class="button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm dialog-icon-info hover:brightness-90 dark:hover:brightness-150 bg-[color:var(--info-button-color)] text-white grow"
        type="button"
      >
        Expand All
      </button>
      <button
        class="button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm dialog-icon-info hover:brightness-90 dark:hover:brightness-150 bg-[color:var(--info-button-color)] text-white grow"
        type="button"
      >
        Collapse All
      </button>
    </div>
  </aside>
</DocumentFragment>
`;

exports[`AppResourcesAside (expanded case) expanded case 2`] = `
<DocumentFragment>
  <aside
    class="
        !gap-2 sm:overflow-visible
        flex flex-col gap-4 overflow-scroll
  overflow-x-auto [overflow-y:overlay] [scrollbar-gutter:auto] 
      "
  >
    <ul
      class=" flex flex-1 flex-col gap-1 overflow-auto"
      role="tree"
    >
      <li
        aria-expanded="true"
        aria-labelledby="app-resources-tree-9-label"
        class="flex flex-col gap-2"
        id="app-resources-tree-9-li"
        role="treeitem"
      >
        <button
          aria-controls="app-resources-tree-9-li"
          class="link inline text-left font-bold"
          id="app-resources-tree-9-label"
          type="button"
        >
          Global Resources 
          <span
            class="pl-2 text-neutral-500"
          >
            (2)
          </span>
        </button>
        <ul
          aria-label="Resources"
          class=" pl-4"
          role="group"
        >
          <li>
            <a
              class="link [&:not([aria-current]):not(:hover)]:!text-neutral-500"
              href="/specify/resources/app-resource/1/"
            >
              <span
                aria-hidden="true"
                title="Other Properties Resource"
              >
                <svg
                  aria-hidden="true"
                  class="w-6 h-6 flex-shrink-0"
                  fill="currentColor"
                  viewBox="0 0 20 20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M3 4a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zm0 4a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zm0 4a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zm0 4a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              Global Preferences
            </a>
          </li>
          <li>
            <a
              class="link [&:not([aria-current]):not(:hover)]:!text-neutral-500"
              href="/specify/resources/app-resource/2/"
            >
              <span
                aria-hidden="true"
                title="Other Properties Resource"
              >
                <svg
                  aria-hidden="true"
                  class="w-6 h-6 flex-shrink-0"
                  fill="currentColor"
                  viewBox="0 0 20 20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M3 4a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zm0 4a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zm0 4a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zm0 4a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              Remote Preferences
            </a>
          </li>
          <li>
            <a
              class="link"
              href="/specify/resources/create/globalResource/"
            >
              <span
                class="!text-green-700 print:hidden"
              >
                <svg
                  aria-hidden="true"
                  class="w-6 h-6 flex-shrink-0"
                  fill="currentColor"
                  viewBox="0 0 20 20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              Add Resource
            </a>
          </li>
        </ul>
      </li>
      <li
        aria-expanded="true"
        aria-labelledby="app-resources-tree-10-label"
        class="flex flex-col gap-2"
        id="app-resources-tree-10-li"
        role="treeitem"
      >
        <button
          aria-controls="app-resources-tree-10-li"
          class="link inline text-left font-bold"
          id="app-resources-tree-10-label"
          type="button"
        >
          Discipline Resources 
          <span
            class="pl-2 text-neutral-500"
          >
            (4)
          </span>
        </button>
        <ul
          aria-label="Sub-categories"
          class=" flex flex-col gap-2 pl-4"
          role="group"
        >
          <li
            aria-expanded="true"
            aria-labelledby="app-resources-tree-11-label"
            class="flex flex-col gap-2"
            id="app-resources-tree-11-li"
            role="treeitem"
          >
            <button
              aria-controls="app-resources-tree-11-li"
              class="link inline text-left font-bold"
              id="app-resources-tree-11-label"
              type="button"
            >
              Botany 
              <span
                class="pl-2 text-neutral-500"
              >
                (4)
              </span>
            </button>
            <ul
              aria-label="Resources"
              class=" pl-4"
              role="group"
            >
              <li>
                <a
                  class="link"
                  href="/specify/resources/create/discipline_3/"
                >
                  <span
                    class="!text-green-700 print:hidden"
                  >
                    <svg
                      aria-hidden="true"
                      class="w-6 h-6 flex-shrink-0"
                      fill="currentColor"
                      viewBox="0 0 20 20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                  Add Resource
                </a>
              </li>
            </ul>
            <ul
              aria-label="Sub-categories"
              class=" flex flex-col gap-2 pl-4"
              role="group"
            >
              <li
                aria-expanded="true"
                aria-labelledby="app-resources-tree-12-label"
                class="flex flex-col gap-2"
                id="app-resources-tree-12-li"
                role="treeitem"
              >
                <button
                  aria-controls="app-resources-tree-12-li"
                  class="link inline text-left font-bold"
                  id="app-resources-tree-12-label"
                  type="button"
                >
                  c 
                  <span
                    class="pl-2 text-neutral-500"
                  >
                    (4)
                  </span>
                </button>
                <ul
                  aria-label="Resources"
                  class=" pl-4"
                  role="group"
                >
                  <li>
                    <a
                      class="link [&:not([aria-current]):not(:hover)]:!text-neutral-500"
                      href="/specify/resources/app-resource/6/"
                    >
                      <span
                        aria-hidden="true"
                        title="Collection Preferences"
                      >
                        <svg
                          aria-hidden="true"
                          class="w-6 h-6 flex-shrink-0"
                          fill="currentColor"
                          viewBox="0 0 20 20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            clip-rule="evenodd"
                            d="M11.49 3.17c-.38-1.56-2.6-1.56-2.98 0a1.532 1.532 0 01-2.286.948c-1.372-.836-2.942.734-2.106 2.106.54.886.061 2.042-.947 2.287-1.561.379-1.561 2.6 0 2.978a1.532 1.532 0 01.947 2.287c-.836 1.372.734 2.942 2.106 2.106a1.532 1.532 0 012.287.947c.379 1.561 2.6 1.561 2.978 0a1.533 1.533 0 012.287-.947c1.372.836 2.942-.734 2.106-2.106a1.533 1.533 0 01.947-2.287c1.561-.379 1.561-2.6 0-2.978a1.532 1.532 0 01-.947-2.287c.836-1.372-.734-2.942-2.106-2.106a1.532 1.532 0 01-2.287-.947zM10 13a3 3 0 100-6 3 3 0 000 6z"
                            fill-rule="evenodd"
                          />
                        </svg>
                      </span>
                      Collection Preferences
                    </a>
                  </li>
                  <li>
                    <a
                      class="link"
                      href="/specify/resources/create/collection_4/"
                    >
                      <span
                        class="!text-green-700 print:hidden"
                      >
                        <svg
                          aria-hidden="true"
                          class="w-6 h-6 flex-shrink-0"
                          fill="currentColor"
                          viewBox="0 0 20 20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            clip-rule="evenodd"
                            d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z"
                            fill-rule="evenodd"
                          />
                        </svg>
                      </span>
                      Add Resource
                    </a>
                  </li>
                </ul>
                <ul
                  aria-label="Sub-categories"
                  class=" flex flex-col gap-2 pl-4"
                  role="group"
                >
                  <li
                    aria-expanded="true"
                    aria-labelledby="app-resources-tree-13-label"
                    class="flex flex-col gap-2"
                    id="app-resources-tree-13-li"
                    role="treeitem"
                  >
                    <button
                      aria-controls="app-resources-tree-13-li"
                      class="link inline text-left font-bold"
                      id="app-resources-tree-13-label"
                      type="button"
                    >
                      User Accounts 
                      <span
                        class="pl-2 text-neutral-500"
                      >
                        (3)
                      </span>
                    </button>
                    <ul
                      aria-label="Sub-categories"
                      class=" flex flex-col gap-2 pl-4"
                      role="group"
                    >
                      <li
                        aria-expanded="true"
                        aria-labelledby="app-resources-tree-14-label"
                        class="flex flex-col gap-2"
                        id="app-resources-tree-14-li"
                        role="treeitem"
                      >
                        <button
                          aria-controls="app-resources-tree-14-li"
                          class="link inline text-left font-bold"
                          id="app-resources-tree-14-label"
                          type="button"
                        >
                          testiiif 
                          <span
                            class="pl-2 text-neutral-500"
                          >
                            (3)
                          </span>
                        </button>
                        <ul
                          aria-label="Resources"
                          class=" pl-4"
                          role="group"
                        >
                          <li>
                            <a
                              class="link [&:not([aria-current]):not(:hover)]:!text-neutral-500"
                              href="/specify/resources/app-resource/5/"
                            >
                              <span
                                aria-hidden="true"
                                title="User Preferences"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="w-6 h-6 flex-shrink-0"
                                  fill="currentColor"
                                  viewBox="0 0 20 20"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    clip-rule="evenodd"
                                    d="M11.49 3.17c-.38-1.56-2.6-1.56-2.98 0a1.532 1.532 0 01-2.286.948c-1.372-.836-2.942.734-2.106 2.106.54.886.061 2.042-.947 2.287-1.561.379-1.561 2.6 0 2.978a1.532 1.532 0 01.947 2.287c-.836 1.372.734 2.942 2.106 2.106a1.532 1.532 0 012.287.947c.379 1.561 2.6 1.561 2.978 0a1.533 1.533 0 012.287-.947c1.372.836 2.942-.734 2.106-2.106a1.533 1.533 0 01.947-2.287c1.561-.379 1.561-2.6 0-2.978a1.532 1.532 0 01-.947-2.287c.836-1.372-.734-2.942-2.106-2.106a1.532 1.532 0 01-2.287-.947zM10 13a3 3 0 100-6 3 3 0 000 6z"
                                    fill-rule="evenodd"
                                  />
                                </svg>
                              </span>
                              User Preferences
                            </a>
                          </li>
                          <li>
                            <a
                              class="link [&:not([aria-current]):not(:hover)]:!text-neutral-500"
                              href="/specify/resources/app-resource/4/"
                            >
                              <span
                                aria-hidden="true"
                                title="Other XML Resource"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="w-6 h-6 flex-shrink-0"
                                  fill="currentColor"
                                  viewBox="0 0 20 20"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    clip-rule="evenodd"
                                    d="M12.316 3.051a1 1 0 01.633 1.265l-4 12a1 1 0 11-1.898-.632l4-12a1 1 0 011.265-.633zM5.707 6.293a1 1 0 010 1.414L3.414 10l2.293 2.293a1 1 0 11-1.414 1.414l-3-3a1 1 0 010-1.414l3-3a1 1 0 011.414 0zm8.586 0a1 1 0 011.414 0l3 3a1 1 0 010 1.414l-3 3a1 1 0 11-1.414-1.414L16.586 10l-2.293-2.293a1 1 0 010-1.414z"
                                    fill-rule="evenodd"
                                  />
                                </svg>
                              </span>
                              QueryExtraList
                            </a>
                          </li>
                          <li>
                            <a
                              class="link [&:not([aria-current]):not(:hover)]:!text-neutral-500"
                              href="/specify/resources/app-resource/3/"
                            >
                              <span
                                aria-hidden="true"
                                title="Other XML Resource"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="w-6 h-6 flex-shrink-0"
                                  fill="currentColor"
                                  viewBox="0 0 20 20"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    clip-rule="evenodd"
                                    d="M12.316 3.051a1 1 0 01.633 1.265l-4 12a1 1 0 11-1.898-.632l4-12a1 1 0 011.265-.633zM5.707 6.293a1 1 0 010 1.414L3.414 10l2.293 2.293a1 1 0 11-1.414 1.414l-3-3a1 1 0 010-1.414l3-3a1 1 0 011.414 0zm8.586 0a1 1 0 011.414 0l3 3a1 1 0 010 1.414l-3 3a1 1 0 11-1.414-1.414L16.586 10l-2.293-2.293a1 1 0 010-1.414z"
                                    fill-rule="evenodd"
                                  />
                                </svg>
                              </span>
                              QueryFreqList
                            </a>
                          </li>
                          <li>
                            <a
                              class="link"
                              href="/specify/resources/create/collection_4_user_1/"
                            >
                              <span
                                class="!text-green-700 print:hidden"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="w-6 h-6 flex-shrink-0"
                                  fill="currentColor"
                                  viewBox="0 0 20 20"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    clip-rule="evenodd"
                                    d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z"
                                    fill-rule="evenodd"
                                  />
                                </svg>
                              </span>
                              Add Resource
                            </a>
                          </li>
                        </ul>
                      </li>
                    </ul>
                  </li>
                  <li
                    aria-expanded="true"
                    aria-labelledby="app-resources-tree-15-label"
                    class="flex flex-col gap-2"
                    id="app-resources-tree-15-li"
                    role="treeitem"
                  >
                    <button
                      aria-controls="app-resources-tree-15-li"
                      class="link inline text-left font-bold"
                      id="app-resources-tree-15-label"
                      type="button"
                    >
                      User Types 
                      <span
                        class="pl-2 text-neutral-500"
                      >
                        (0)
                      </span>
                    </button>
                    <ul
                      aria-label="Sub-categories"
                      class=" flex flex-col gap-2 pl-4"
                      role="group"
                    >
                      <li
                        aria-expanded="false"
                        aria-labelledby="app-resources-tree-16-label"
                        class="flex flex-col gap-2"
                        id="app-resources-tree-16-li"
                        role="treeitem"
                      >
                        <button
                          aria-controls="app-resources-tree-16-li"
                          class="link inline text-left font-bold"
                          id="app-resources-tree-16-label"
                          type="button"
                        >
                          FullAccess 
                          <span
                            class="pl-2 text-neutral-500"
                          >
                            (0)
                          </span>
                        </button>
                      </li>
                      <li
                        aria-expanded="false"
                        aria-labelledby="app-resources-tree-17-label"
                        class="flex flex-col gap-2"
                        id="app-resources-tree-17-li"
                        role="treeitem"
                      >
                        <button
                          aria-controls="app-resources-tree-17-li"
                          class="link inline text-left font-bold"
                          id="app-resources-tree-17-label"
                          type="button"
                        >
                          Guest 
                          <span
                            class="pl-2 text-neutral-500"
                          >
                            (0)
                          </span>
                        </button>
                      </li>
                      <li
                        aria-expanded="false"
                        aria-labelledby="app-resources-tree-18-label"
                        class="flex flex-col gap-2"
                        id="app-resources-tree-18-li"
                        role="treeitem"
                      >
                        <button
                          aria-controls="app-resources-tree-18-li"
                          class="link inline text-left font-bold"
                          id="app-resources-tree-18-label"
                          type="button"
                        >
                          LimitedAccess 
                          <span
                            class="pl-2 text-neutral-500"
                          >
                            (0)
                          </span>
                        </button>
                      </li>
                      <li
                        aria-expanded="false"
                        aria-labelledby="app-resources-tree-19-label"
                        class="flex flex-col gap-2"
                        id="app-resources-tree-19-li"
                        role="treeitem"
                      >
                        <button
                          aria-controls="app-resources-tree-19-li"
                          class="link inline text-left font-bold"
                          id="app-resources-tree-19-label"
                          type="button"
                        >
                          Manager 
                          <span
                            class="pl-2 text-neutral-500"
                          >
                            (0)
                          </span>
                        </button>
                      </li>
                    </ul>
                  </li>
                </ul>
              </li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
    <div
      class="flex flex-wrap gap-2"
    >
      <button
        class="button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm dialog-icon-info hover:brightness-90 dark:hover:brightness-150 bg-[color:var(--info-button-color)] text-white grow"
        type="button"
      >
        Expand All
      </button>
      <button
        class="button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm dialog-icon-info hover:brightness-90 dark:hover:brightness-150 bg-[color:var(--info-button-color)] text-white grow"
        type="button"
      >
        Collapse All
      </button>
    </div>
  </aside>
</DocumentFragment>
`;

exports[`AppResourcesAside (simple no conformation case) simple no conformation case 1`] = `
<DocumentFragment>
  <aside
    class="
        !gap-2 sm:overflow-visible
        flex flex-col gap-4 overflow-scroll
  overflow-x-auto [overflow-y:overlay] [scrollbar-gutter:auto] 
      "
  >
    <ul
      class=" flex flex-1 flex-col gap-1 overflow-auto"
      role="tree"
    >
      <li
        aria-expanded="false"
        aria-labelledby="app-resources-tree-0-label"
        class="flex flex-col gap-2"
        id="app-resources-tree-0-li"
        role="treeitem"
      >
        <button
          aria-controls="app-resources-tree-0-li"
          class="link inline text-left font-bold"
          id="app-resources-tree-0-label"
          type="button"
        >
          Global Resources 
          <span
            class="pl-2 text-neutral-500"
          >
            (2)
          </span>
        </button>
      </li>
      <li
        aria-expanded="false"
        aria-labelledby="app-resources-tree-1-label"
        class="flex flex-col gap-2"
        id="app-resources-tree-1-li"
        role="treeitem"
      >
        <button
          aria-controls="app-resources-tree-1-li"
          class="link inline text-left font-bold"
          id="app-resources-tree-1-label"
          type="button"
        >
          Discipline Resources 
          <span
            class="pl-2 text-neutral-500"
          >
            (4)
          </span>
        </button>
      </li>
    </ul>
    <div
      class="flex flex-wrap gap-2"
    >
      <button
        class="button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm dialog-icon-info hover:brightness-90 dark:hover:brightness-150 bg-[color:var(--info-button-color)] text-white grow"
        type="button"
      >
        Expand All
      </button>
      <button
        class="button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm dialog-icon-info hover:brightness-90 dark:hover:brightness-150 bg-[color:var(--info-button-color)] text-white grow"
        type="button"
      >
        Collapse All
      </button>
    </div>
  </aside>
</DocumentFragment>
`;
